### ISSUE TITLE
How to visualize the network output?

### ISSUE THREAD

**Issue by farazkhan86** [wolny/pytorch-3dunet#7]
Hi,

As required by your code, I have now saved all my volumes into ``hdf5`` files with the dataset name ``raw`` for the image and ``label`` for its corresponding target.
The network has been running for over a day now and is reporting very good accuracy, I would like to see how it is performing by exporting the 3D volume as a ``.tif`` file.

I've written a file ``visualize_output.py`` to do just that and used the best checkpoint pytorch file. I reused most of your code in there to get what I want.

The problem is that I am unable to visualize the output as the output from the network comes back as black volumes. As if every predicted output pixel is ``0``. I have tried with every possible ``datatype`` but still get same black volume. Is the way I am trying to extract my data wrong?

`predicted = torch.argmax(output.detach(), 1).squeeze(0)
`

I'd like to note that the ``target`` volumes are being written just fine and I can see those by opening the ``tif``` files.

Here is my code for visualized a predicted output: https://pastebin.com/Q6AHeJkD

Thank you

**Comment by farazkhan86** [comment]
@wolny 

Just to post an update, rather than using my own function (which was clearly not working to show me the output), I used the ``predict.py`` function and used the saved ``hdf5`` file to extract the ``probability_map``, but that also did not show me the output. Here is what I did:

```
path = 'model_checkpoints\\probabilities.h5'
hdf5_file = h5py.File(path, 'r')
volume = hdf5_file['probability_maps']
data = np.array(volume[:,:,:])
io.volwrite('out_file.tif', data)
```

The output file produced the following:


![out_file](https://user-images.githubusercontent.com/44167893/49584725-146c2180-f954-11e8-9c7d-1cc89e72b08e.jpg)


Its of the correct dimension: ``21*512*1024`` but I cant see the output.

Was the probability_map output from ``predict.py`` meant to be visualized?

Thanks as always

**Comment by farazkhan86** [comment]
Sorry but another update.

 I loaded up the tensorboard logs after the training finished and I got an accuracy of ``0.999`` and the distributions show just that, but under the images tab I get the following:

![screenshot_2018-12-06_14-28-59](https://user-images.githubusercontent.com/44167893/49590468-e42c7f00-f963-11e8-8824-d13fa58ad6f5.png)
![screenshot_2018-12-06_14-29-35](https://user-images.githubusercontent.com/44167893/49590470-e42c7f00-f963-11e8-8a2e-46337924b304.png)
![screenshot_2018-12-06_14-30-00](https://user-images.githubusercontent.com/44167893/49590471-e42c7f00-f963-11e8-9c55-a01f04303e0e.png)

Why is the target coming up as black as well? Is that the reason I am getting nothing on prediction?

When I check manually the target is a binary file which represent those white spots/blobs shown in the input image.

Thanks

**Comment by farazkhan86** [comment]
Another update,

I believe I may have figured out why my ``output`` from network always has values ``[[[0.997, 0.989, ...]]]``` and also why ``predict.py`` may have produced the all white output when saved as output to see.

The ``_expand_target()`` function was changing my label tensor. My label tensors were already ready as a binary tensor, with background as ``0`` and the small points of interest as ``1`` and were in shape ``[DxHxW]``, but the following code in ``_expand_target()`` would change the label tensor to all ``1``.

```
     result = torch.zeros(shape)
     # for each batch instance
     for i in range(input.size()[0]):
         # iterate over channel axis and create corresponding binary mask in the target
         for c in range(C):
             mask = result[i, c]
             mask[input[i] == c] = 1
```
I have replaced this with a simple: ``input.unsqueeze(1)``. This keeps the label tensor as expected and doesnt change the binary values.

I've re-run the system for overnight training and can only hope that this was the reason the network was not performing well :(

**Comment by wolny** [comment]
Hi @farazkhan86,
how did the training go?
Just couple of points:
1. The output from the network is always 4D tensor: `CxDxHxW`, where `C` is the number of classes/channels. So your code above:
```
volume = hdf5_file['probability_maps']
data = np.array(volume[:,:,:])
```
will not produce correct results.
2. The `_expand_target` function is used for computing the Dice Score (accuracy) as well as tensorboard visualization. Usually your target is just 3D tensor where each voxel contains a label value (`0,...,C-1`, where `C` is the number of classes). If you're training with `CrossEntropyLoss` or `WeightedCrossEntropyLoss`, that's what the loss function requires. However for computing the Dice Score it's needed expand the target dimensionality and put the classes in separate channels, so that it matches the dimensionality of the `output`.
3. In case of prediction: there was a *bug* in the `predict.py` for networks trained with `CE` and `WCE` losses. Basically the outputs were not normalized with the final activation function (`Softmax` in this case). This might have been the root cause of your problems if you've trained with `CE` or `WCE`. The problem is fixed in the latest master so don't forget to pull it when running your experiment.

**Comment by farazkhan86** [comment]
Hi @wolny ,

Thanks for replying. Unfortunately I have still not been able to resolve this and I am struggling to get any meaningful output using ``diceloss``

> The output from the network is always 4D tensor: CxDxHxW, where C is the number of classes/channels. So your code above:
> volume = hdf5_file['probability_maps']
> data = np.array(volume[:,:,:])
> will not produce correct results.

So what would be the correct way to visualize the final prediction of the network?

> The _expand_target function is used for computing the Dice Score (accuracy) as well as tensorboard visualization. Usually your target is just 3D tensor where each voxel contains a label value (0,...,C-1, where C is the number of classes). If you're training with CrossEntropyLoss or WeightedCrossEntropyLoss, that's what the loss function requires. However for computing the Dice Score it's needed expand the target dimensionality and put the classes in separate channels, so that it matches the dimensionality of the output.

I understand, so given my data, where the target tensor is a binary tensor of shape ``[DxHxW]``, the points of interest represented by 1 and background by 0. Should I change the output of the network with ``out_channels = 2`` as there are effectively two classes of background and foreground? I'm trying to understand how to handle a simple problem of background and foreground with diceloss..

> In case of prediction: there was a bug in the predict.py for networks trained with CE and WCE losses. Basically the outputs were not normalized with the final activation function (Softmax in this case). This might have been the root cause of your problems if you've trained with CE or WCE. The problem is fixed in the latest master so don't forget to pull it when running your experiment.

Thanks for fixing the bug, but I have been using ``diceloss``. I think the problem I am facing is completely related to point 2, where either my output channels from the network are not correct or my target tensor is not.

Many thanks

**Comment by wolny** [comment]
Hi @farazkhan86, sorry for the late response. I thought I replied to you already. There was indeed an issue when using the predict script with the network that was trained with `dice` loss and a single output channel. I was playing around with some dataset coming from the confocal microscope recently where the membrane of the cells was stained, and the ground truth was a binary mask (membrane/background) similar to your case. I was able to train the network successfully with the `dice` loss and a single output channel, but when running a predict on the test data the output was just black. The reason was that the network was using `Softmax` instead of `Sigmoid` normalization in the predict script.
In order to train with standard `dice` loss and a single binary target:
```
python train.py --checkpoint-dir ./logs --in-channels 1 --out-channels 1 --validate-after-iters 50 --log-after-iters 50 --interpolate --loss bce --train-path <train-dataset> --val-path <val-dataset> --train-patch <patch-size> --train-stride <train-stride> --val-patch <val-patch> --val-stride <val-stride> --transformer <transformer> --final-sigmoid
```
notice the *--final-sigmoid* passed as an option. (please make sure to checkout the latest master first).
Then, in order to predict on the test dataset:
```
python predict.py --model-path <model-path> --in-channels 1 --out-channels 1 --interpolate --test-path <test-path>  --patch <patch> --stride <stride> --final-sigmoid
```
Hope that helps.

**Comment by farazkhan86** [comment]
@wolny 

Thanks for finally replying. :)

Thats great to know that you were able to train a network with good results on a dataset similar to mine. All hope may not be lost for my case :)

>  I was able to train the network successfully with the dice loss and a single output channel

Can you please clarify this a bit. You say that your dataset was similar as you had background and foreground as your label tensors. Based on what you said in your earlier reply, which I have been following:

> However for computing the Dice Score it's needed expand the target dimensionality and put the classes in separate channels, so that it matches the dimensionality of the output.

So I thought that with such a data, the network output has to be 2 channels; one for background and one for foreground. But since you got good results with 1 channel, I'll try it again as well. Does that mean that I don't have to use ``expand target`` even though I am using ``diceLoss``?

P.S. Since you also worked with confocal images, may i know how you are loading them in? My samples are in ``.lif`` format and the only way I can load them in is as stacked ``.tiff`` files as ``simpleITK`` and other mri libraries do not support this image format.

Thanks

**Comment by wolny** [comment]
> Can you please clarify this a bit. You say that your dataset was similar as you had background and foreground as your label tensors...
yeah sorry for the confusion, you should totally be able to train the network with `DiceLoss` or `BinaryCrossEntropy` without splitting your target in separate channels. The splitting would be necessary if you train with `CrossEntropyLoss`.

> Since you also worked with confocal images, may i know how you are loading them in? My samples are in .lif format and the only way I can load them in is as stacked .tiff files as simpleITK and other mri libraries do not support this image format.

If you have them in `.tiff` format the best way is to open your stacks in Fiji and export as `HDF5`. I'm using `ilastik` plugin in order to convert `.tiff` to `HDF5` (https://github.com/ilastik/ilastik4ij).

**Comment by farazkhan86** [comment]
> yeah sorry for the confusion, you should totally be able to train the network with DiceLoss or BinaryCrossEntropy without splitting your target in separate channels. The splitting would be necessary if you train with CrossEntropyLoss.

Thats good to know, thanks for the clarification. I'll give it another go with the suggested settings and let you know how I get on.

> If you have them in .tiff format the best way is to open your stacks in Fiji and export as HDF5. I'm using ilastik plugin in order to convert .tiff to HDF5 (https://github.com/ilastik/ilastik4ij).

``tiff`` is not the format I have my samples in. They are in ``.lif`` and there was no way to directly load them into python, as the common mri libraries and simpleITK failed to load them in, so I then exported the samples as tiff files using ImageJ.

**Comment by farazkhan86** [comment]
Hi @wolny ,

Although you were successful, I was unable to recreate your results using the same single channel method on my binary data.

I would really like this to work as this is frustrating since you mentioned you trained your network using data similar to mine using a single channel and dice loss, but I dont know how that is possible as the function ``expand_target`` in the ``losses.py`` file would distort the target for a single channel.

As mentioned I have a confocal microscope dataset and for every input I have a binary mask as the target, where the background are represented by ``0`` and the membrane by ``1``. 

Now such a tensor when given to ``expand_target``:  
```
 for i in range(input.size()[0]):
        # iterate over channel axis and create corresponding binary mask in the target
        for c in range(C):
            mask = result[i, c]
            mask[input[i] == c] = 1
```

would basically change the target tensor/ binary mask entirely by changing the background bits to ``1``. And after manually checking the function, for a single channel the function indeed changes the mask and thus changes the target tensor. So from the first iteration we get accuracies of ``99%``.

Is this function supposed to work like this? How did you train this network on similar data as mine using dice and single output channel?


After a full weekend of trying and overnight training, I could not get any meaningful result, as can be seen from the ``tensorboard`` screenshots

![1](https://user-images.githubusercontent.com/44167893/51496403-8c34c500-1db7-11e9-9a52-8c8e0d2d4b4e.PNG)
![2](https://user-images.githubusercontent.com/44167893/51496409-91920f80-1db7-11e9-9cf4-b360e41f6ce7.PNG)


Just for reference, attached is a single sample, its a hdf5 file written as suggested in the readme with the data in ``raw`` and target in ``label``.  Given such a simple dataset, why is this network failing to give any results?
Thanks
[volume_0_1.zip](https://github.com/wolny/pytorch-3dunet/files/2780385/volume_0_1.zip)

**Comment by wolny** [comment]
Hi @farazkhan86,

I'm surprised that `expand_target` is called at all in your case as it shouldn't. It's only meant to be called in multi-class segmentation scenario when the target tensor is expected to be 4D (NDHW) with values belonging to `[0, C-1]`, where `C` is the number of classes (see the `if` statement: https://github.com/wolny/pytorch-3dunet/blob/master/unet3d/losses.py#L13).

So the reason it works for me and it doesn't for you is that the `target` patch returned from my loader is 5D (NCDHW) with channel dimension of size 1, instead of 4D (NDHW) as in your case. 

TL;DR you have to make sure that the channel dimension is there in your target, so that `expand_target` is not called (btw the function is called `expand_as_one_hot` in the latest master).
Which `Transformer` are you using for your dataset? Is it https://github.com/wolny/pytorch-3dunet/blob/master/augment/transforms.py#L259? Then change `https://github.com/wolny/pytorch-3dunet/blob/master/augment/transforms.py#L279` to `expand_dims=True` and try again.
Alternatively you could:
1. Just add `target = torch.unsqueeze(target, 1)` to transform from 4D to 5D somewhere in the code (e.g. after `ToTensor` transformation)
1. Expand raw and label datasets in your H5 to contain the channel dimension of size 1 (axis order should be CDHW)

This might indeed be confusing, so I'll mention it in the README and ideally add some constraints in the code which prevent such errors.

**Comment by farazkhan86** [comment]
Thanks for replying.

Yes as I expected from the beginning, ``expand_target`` was the source of all my problems and the reason for me getting instant 99% accuracies yest poor segmentation results. 

Thank you for your suggestion in changing either the data loader or the transforms function. I managed to achieve similar results by inverting my ``label`` values so that when it goes through ``expand_target`` it is output in the proper and expected form.

Thank you :)
